src/test/java/engineering/swat/watch/RecursiveWatchTests.java

```diff
@@ -0,0 +1,120 @@
+package engineering.swat.watch;
+
+import static org.awaitility.Awaitility.await;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import engineering.swat.watch.WatchEvent.Kind;
+
+class RecursiveWatchTests {
+    private final Logger logger = LogManager.getLogger();
+
+    private TestDirectory testDir;
+
+    @BeforeEach
+    void setup() throws IOException {
+        testDir = new TestDirectory();
+    }
+
+    @AfterEach
+    void cleanup() throws IOException {
+        if (testDir != null) {
+            testDir.close();
+        }
+    }
+
+    @BeforeAll
+    static void setupEverything() throws IOException {
+        Awaitility.setDefaultTimeout(TestHelper.NORMAL_WAIT);
+    }
+
+
+    @Test
+    void newDirectoryWithFilesChangesDetected() throws IOException {
+        var target = new AtomicReference<Path>();
+        var created = new AtomicBoolean(false);
+        var changed = new AtomicBoolean(false);
+        var watchConfig = Watcher.watch(testDir.getTestDirectory(), WatchScope.PATH_AND_ALL_DESCENDANTS)
+            .onEvent(ev -> {
+                    logger.debug("Event received: {}", ev);
+                    if (ev.calculateFullPath().equals(target.get())) {
+                        switch (ev.getKind()) {
+                            case CREATED:
+                                created.set(true);
```

Comment:
Could this have the same issue as one of the issues we experienced yesterday? (The `CREATED` event being missed, and getting a `MODIFIED` event instead.) Maybe the probability of it going wrong without "torturing" the system is very low, but maybe it's still worth the effort to fix a bit of unlikely-but-still-possible flakiness.

Reply:
well, this test hasn't failed yet. and indeed, it only goes wrong when you really hammer the system.

and I like this, as it also would show the negatives of a alternative implementation, it should break on this.